packages/plugin-cloudflare/src/tunnel.ts

```diff
@@ -45,7 +47,16 @@ class TunnelClientInstance implements TunnelClient {
 
   constructor(port: number) {
     this.port = port
-    this.tunnel()
+  }
+
+  async startTunnel() {
+    try {
+      await install()
```

Comment:
I am a JS newbie, just curious why the change away from the callback structure?

Reply:
I only used then/catch initially because you can't use `await` inside a constructor, but we prefer to use `async/await` in our codebase because it's easier to read and debug. (is not a good practice to use both approaches in a codebase)

But after looking at the implementation when adding tests, I thought it made sense to actually move that from the constructor into a "setup" function, so that the constructor doesn't have side effects (only builds the object)

This way is also easier to test.

Reply:
Those are valid questions! 

> If anyone wants to interact with an instance of TunnelClientInstance via the TunnelClient interface, they may not have a functional tunnel because the tunnel must be started manually via startTunnel() which is not available on the TunnelClient interface

This is actually the intention, but let me explain. Consumers of this interface won't create the instance themselves, they run a hook called `tunnel_start` that calls [`hookStart`](https://github.com/Shopify/cli/blob/main/packages/plugin-cloudflare/src/tunnel.ts#L28), that function is the one that crates (and starts!) the tunnel and returns the instance to control the tunnel (TunnelClient). This is because the tunnel lives in a different package and you could potentially have multiple packages that implement the `tunnel_start` hook.

Hooks are an Oclif thing, more about them [here](https://oclif.io/docs/hooks/))


So from the point of view of the `app` package (the consumer of this interface), it [executes](https://github.com/Shopify/cli/blob/main/packages/app/src/cli/services/dev/urls.ts#L288) a `tunnel_start` hook and receives in return a `TunnelClient` with the tunnel already started :)